engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/TypeOfNode.java

```
@@ -0,0 +1,108 @@
+    Context ctx = Context.get(this);
+    return ctx.getBuiltins().dateTime();
+  }
+
+  @Specialization(
+      guards = {"!interop.isDate(value)", "!interop.isTime(value)", "interop.isTimeZone(value)"})
+  Object doTimeZone(Object value, @CachedLibrary(limit = "3") InteropLibrary interop) {
+    Context ctx = Context.get(this);
+    return ctx.getBuiltins().timeZone();
+  }
+
+  @Specialization(guards = {"interop.isDate(value)", "!interop.isTime(value)"})
+  Object doDate(Object value, @CachedLibrary(limit = "3") InteropLibrary interop) {
+    Context ctx = Context.get(this);
+    return ctx.getBuiltins().date();
+  }
+
+  @Specialization(guards = {"!interop.isDate(value)", "interop.isTime(value)"})
+  Object doTime(Object value, @CachedLibrary(limit = "3") InteropLibrary interop) {
+    Context ctx = Context.get(this);
+    return ctx.getBuiltins().timeOfDay();
+  }
+
+  @Specialization(
+      guards = {
+        "interop.hasMetaObject(value)",
+        "!types.hasType(value)",
+        "!interop.isDate(value)",
+        "!interop.isTime(value)",
+        "!interop.isTimeZone(value)"
+      })
+  Object doMetaObject(
+      Object value,
+      @CachedLibrary(limit = "3") InteropLibrary interop,
+      @CachedLibrary(limit = "3") TypesLibrary types) {
+    try {
+      return interop.getMetaObject(value);
+    } catch (UnsupportedMessageException e) {
+      CompilerDirectives.transferToInterpreter();
+      Builtins builtins = Context.get(this).getBuiltins();
+      throw new PanicException(builtins.error().makeCompileError("invalid meta object"), this);
+    }
+  }
+
+  @Specialization(guards = {"types.hasType(value)", "!interop.isNumber(value)"})
+  Object doType(
+      Object value,
+      @CachedLibrary(limit = "3") InteropLibrary interop,
+      @CachedLibrary(limit = "3") TypesLibrary types) {
+    return types.getType(value);
+  }
+
+  @Fallback
+  Object doAny(Object value) {
+    return Context.get(this).getBuiltins().any();
+  }
```

Are you sure this is the way to go?

I understand that if for some reason we cannot figure out the type of some objects (do we have any real-life examples when this happens? or is it an 'just-in-case' branch?) I'd rather report some kind of error "sorry the type is unknown". I know that `anything <: Any`, but here we don't want to return any supertype but the most specific one and in this case `Any` is not most specific, we just cannot resolve the correct one. My intuition is it is better to fail clearly than do silent implicit fallbacks like this one. But that is of course just an opinion and very open to discussion.

---

I will always take some default top type rather than having a crash that manifests itself to the user. But that is indeed my preference. 

I can't really come up with any example that would hit this case atm.

---

It isn't really a crash, just an error saying that we couldn't find the thing.

I think we have this philosophy to prefer a crash over data corruption. If returning the more general type is a case of 'data corruption' is probably arguable though.

Personally I don't like that this may allow errors to go unnoticed.

Especially if you don't know any example when this happens - IMO it's better to indeed expose a crash that manifests to the user instead of silently falling back. If they notice the crash, they can still handle it (`Panic.catch`), but they at least know that it happened and can tell us to fix the implementation to handle the missing cases. If they get `Any` they will not know anything unexpected happen and may learn much later when something doesn't work. Maybe this is not a very big issue as `type_of` is probably going to be mostly used by the more advanced users anyway.